### ISSUE TITLE
GitHub avatars: support enterprise avatars with GITHUB_TOKEN, handle failure better

### ISSUE THREAD

**Issue by vuon9** [zed-industries/zed#19207]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Zed made multiple unsuccessful attempts to retrieve the avatar from my private GH enterprise account. I believe it would be better if the retry attempts were limited, and having a default avatar type, such as using the first characters of the first and last names as a fallback, would be ideal.

Zed log:

<img width="1281" alt="image" src="https://github.com/user-attachments/assets/cb9779bc-646c-4281-a3b8-af443dab22b4">

My avatar is blank:
<img width="216" alt="image" src="https://github.com/user-attachments/assets/d769e949-e06f-4a1f-80dd-07b1c95718ce">

### Environment

Zed: v0.156.1 (Zed)
OS: macOS 14.6.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="372" alt="image" src="https://github.com/user-attachments/assets/a165beb4-e9d2-4af0-8446-69c863747bce">

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by notpeter** [comment]
Yeah, I ran into this previously and just dug up the slack conversation.  It appears that when a user is inside a GitHub hosted Enterprise (not surely exactly which plans/settings trigger this) the GitHub API behaves very differently for these users.

For example to fetch my avatar, we would hit: https://api.github.com/users/notpeter which returns a nice json object including an `avatar_url` of https://avatars.githubusercontent.com/u/145113?v=4.

Where as for a [random gh hosted enterprise user](https://api.github.com/users/Alex-Longley_lmig) it just straight 404s.  Similarly, if you hit the [profile link](https://github.com/Alex-Longley_lmig) it redirects you to an enterprise login. Wild.

We absolutely should handle this better (stop retrying after a couple fails), but I'm curious whether you have a `GITHUB_TOKEN` configured (e.g. for the `gh` cli with `export GITHUB_TOKEN="$(gh auth token)"` because ideally we could actually make Avatars within your org not actually display properly if there was a token available.

**Comment by vuon9** [comment]
```
export GITHUB_TOKEN="$(gh auth token)"
```
Indeed, I do have it. I successfully tested the calling profile API using that GH token and receive the same response data structure like public GH account could have.

**Comment by notpeter** [comment]
A user reported a similar issue (repeated failed attempts to load an avatar) for a deleted/renamed github user here:
- https://github.com/zed-industries/zed/issues/19460

We should fail gracefully in that situation as well.

**Comment by mrnugget** [comment]
Two things to do here:

1. Take GITHUB_TOKEN into account for GHE
2. If image fails to load, show placeholder avatar and do not try to re-render

**Comment by benbrandt** [comment]
I am no longer able to reproduce the spamming GitHub issue. It has likely been fixed with the image cache: https://github.com/zed-industries/zed/pull/27774

I hardcoded all avatars to load a specific url I knew would fail, and it is only logging once (I am only seeing one when it fails the first time)

I will keep digging into how we can actually support enterprise images though.

**Comment by notpeter** [comment]
@benbrandt I believe the solution would likely be to use a GITHUB_TOKEN for requests:
- https://github.com/zed-industries/zed/issues/9483

**Comment by benbrandt** [comment]
#32164 would allow for using a GitHub token if provided, however this has a few potential concerns:

1. The user can only do this if they set an env variable and launch Zed from the CLI
2. If the token is expired, we may cause more issues throughout the app, which that PR doesn't handle currently

We could do some form of oauth/refresh token route for the user, but this would be a larger change.

It seems like, no matter what having fallback avatars would be a better experience in the app, because there are other reasons we might fail to load the image. 

We will circle back with @iamnbutler and @danilo-leal on their previous work here and see if we can manage to get some form of avatar fallbacks in the app.

**Comment by notpeter** [comment]
I discovered if you have the github user id (not just the username) you can directly retrieve the avatar image by-id which (1) saves a 302 http redirect (2) works for github enterprise users.

```
# curl -I 'https://github.com/notpeter.png?size=128' |grep location
location: https://avatars.githubusercontent.com/u/145113?s=128&v=4
```

**Comment by vuon9** [comment]
hi @notpeter, the link actually only works if I request within the browser that I have a live session of that GitHub account, it's failed in another browser profile OR just 404 in Zed.

Here is the log I have in Zed.
```
2025-11-03T11:02:44+07:00 ERROR [gpui::asset_cache] Failed to load asset: unexpected http status for https://github.com/your_enterprise_gh_username.png?size=128: 404 Not Found, body: Not Found
```
